ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java

```diff
@@ -85,27 +104,42 @@ private void showSaveError( ClusterSchema clusterSchema, KettleException e ) {
       getMessage( "Spoon.Dialog.ErrorSavingCluster.Message", clusterSchema.getName() ), e );
   }
 
-  public void editClusterSchema( TransMeta transMeta, ClusterSchema clusterSchema ) {
-    ClusterSchemaDialog dialog =
-        new ClusterSchemaDialog( spoon.getShell(), clusterSchema, transMeta.getClusterSchemas(), transMeta.getSlaveServers() );
-    if ( dialog.open() ) {
-      if ( spoon.rep != null && clusterSchema.getObjectId() != null ) {
-        try {
-          saveSharedObjectToRepository( clusterSchema, null );
-        } catch ( KettleException e ) {
-          showSaveError( clusterSchema, e );
+  public void editClusterSchema( ClusterSchemaManagementInterface manager, ClusterSchema clusterSchema ) {
+    String originalName = clusterSchema.getName().trim();
+    try {
+      // execution bowl for listing all slave servers
+      SlaveServerManagementInterface slaveServerManagementInterface =
+        spoon.getExecutionBowl().getManager( SlaveServerManagementInterface.class );
+
+      ClusterSchemaDialog dialog =
+          new ClusterSchemaDialog( spoon.getShell(), clusterSchema, manager.getAll(),
+                                   slaveServerManagementInterface.getAll() );
+      if ( dialog.open() ) {
+        String newName = clusterSchema.getName().trim();
+        manager.add( clusterSchema );
+
+        if ( !newName.equalsIgnoreCase( originalName ) ) {
+          manager.remove( originalName );
+        }
+        if ( spoon.rep != null && clusterSchema.getObjectId() != null ) {
+          try {
+            saveSharedObjectToRepository( clusterSchema, null );
+          } catch ( KettleException e ) {
+            showSaveError( clusterSchema, e );
+          }
         }
+        refreshTree();
       }
-      sharedObjectSyncUtil.synchronizeClusterSchemas( clusterSchema );
-      refreshTree();
+    } catch ( KettleException e ) {
+      new ErrorDialog(
+        spoon.getShell(), BaseMessages.getString( PKG, "Spoon.Dialog.ErrorSavingCluster.Title" ), BaseMessages
+          .getString( PKG, "Spoon.Dialog.ErrorSavingCluster.Message" ), e );
     }
   }
 
-  public void delClusterSchema( TransMeta transMeta, ClusterSchema clusterSchema ) {
+  public void delClusterSchema( ClusterSchemaManagementInterface manager, ClusterSchema clusterSchema ) {
     try {
-
-      int idx = transMeta.getClusterSchemas().indexOf( clusterSchema );
-      transMeta.getClusterSchemas().remove( idx );
+      manager.remove( clusterSchema );
```

Comment:
Is it  prompting for confirmation dialog before deleting

Reply:
It does not appear so. This is the existing behavior, and the story doesn't say it should be any different.